### ISSUE TITLE
Feature request: skew-normal family

### ISSUE THREAD

**Issue by psmorris15** [glmmTMB/glmmTMB#1027]
The current application I am working on requires use of a skew-normal likelihood. The skew-normal family is useful for modeling unimodal responses that lie on the real line but for which the normal distribution is inappropriate due to the presence of skewness. I have added the distribution to my local glmmTMB version following the instructions given in the Hacking glmmTMB vignette.

Would there be any interest in working together to add the skew-normal family to the official glmmTMB release? While the brms R package has implemented the skew-normal, I am not aware of any R packages that allow for its use in a frequentist setting. Such an addition should attract users who need to model their data with a skew-normal likelihood but prefer to use frequentist methods or require a faster runtime than brms can provide.

**Comment by bbolker** [comment]
This seems like a good idea. The easiest way to proceed would probably be for you to make a fork of this repository, post your implementation publicly, then make a pull request.

Were there any surprises/challenges in implementing the SN, or did the vignette give you the information you needed?

**Comment by psmorris15** [comment]
I appreciate the quick reply. For the record, I've submitted this under PR https://github.com/glmmTMB/glmmTMB/pull/1029. 

The vignette had the bulk of the information I needed. A few things: 
- I was not sure how to implement the simulation step for the skew-normal. I didn't have an existing function like rnorm or rgamma to fall back on for this. This is potentially something that can be worked out. 
- I was able to figure out how to get the skew parameter to display on print but not when the summary function is called. 
- In my experience the implementation can be sensitive to the starting values to a larger extent than for the simpler distributions like normal or gamma. The default values often did not work (would get the error "negative log-likelihood is NaN at starting parameter values"), particularly for complex models, requiring me to make reasonable guesses.

**Comment by bbolker** [comment]
Poking around a little bit, this is the R code from `fGarch:::.rsnorm` (which generates a *standardized* skew-normal that then needs to be shifted and scaled).  This "just" needs to be translated into C++ (need to be a little careful with the abs/sign stuff, which is problematic in an autodiff context, but I think  should be OK here; we might need to implement with floating point rather than fancy AD types). See e.g. `rtweedie` in https://github.com/glmmTMB/glmmTMB/blob/master/glmmTMB/src/distrib.h

```r
    weight = xi/(xi + 1/xi)
    z = runif(n, -weight, 1 - weight)
    Xi = xi^sign(z)
    Random = -abs(rnorm(n))/Xi * sign(z)
    m1 = 2/sqrt(2 * pi)
    mu = m1 * (xi - 1/xi)
    sigma = sqrt((1 - m1^2) * (xi^2 + 1/xi^2) + 2 * m1^2 - 1)
    Random = (Random - mu)/sigma
    Random
```

**Comment by bbolker** [comment]
PR merged, but leaving this open for tests and simulation code.

**Comment by psmorris15** [comment]
I've attempted an implementation of a simulation function for the skewnormal based on that given in the `sn::rsn` function. The code is here: (https://github.com/psmorris15/glmmTMB_skewnormal_sim). 

Unfortunately when making draws using the `simulate` function I am observing some strange behavior. Many of the drawn values are duplicated. I've given code for an example below. Out of 10,000 draws there are only 24 unique values. I'm wondering if this is due to some behavior related to the `rnorm` function that I'm not aware of.

Is there anything obvious that I should look into that might be causing this?

`set.seed(50010)

library(sn)
library(glmmTMB)

# simulate data

xi <- 2
omega <- 5
alpha <- -3

n <- 10000

y <- rsn(n = 10000, xi = xi, omega = omega, alpha = alpha)

#obtain transformed parameter values

delta <- alpha/sqrt(1 + alpha^2)
mu <- xi + omega * delta * sqrt(2/pi)
sigma <- sqrt(omega^2 * (1 - 2*delta^2/pi))

# fit model

skewnorm_model <- 
  glmmTMB(formula = y ~ 1,
        family = skewnormal(),
        start = list(beta = -1,
                     betad = 3,
                     psi = -5))

# simulate data from model

sim_test=simulate(skewnorm_model, seed=1)`

**Comment by bbolker** [comment]
The only thing that occurs to me off the top of my head is that [here](https://github.com/psmorris15/glmmTMB_skewnormal_sim/blob/master/glmmTMB/src/distrib.h#L312C1-L312C38) you should probably be using `rnorm()` from TMB, not from the standard library, e.g. as [here](https://github.com/psmorris15/glmmTMB_skewnormal_sim/blob/master/glmmTMB/src/glmmTMB.cpp#L683) (`rnorm(Type(0), Type(1))`) ?

**Comment by psmorris15** [comment]
I'm getting this error from `devtools::document` when I replace with `rnorm(0,1)` with `rnorm(Type(0), Type(1))`:

```
Error in `(function (command = NULL, args = character(), error_on_status = TRUE, …`:
! System command 'R' failed
---
Exit status: 1
stdout & stderr: <printed>
```

**Comment by bbolker** [comment]
Using these two lines works for me:

```
    Type chi = CppAD::abs(rnorm(Type(0), Type(1)));
    Type nrv = rnorm(Type(0), Type(1));
```

**Comment by psmorris15** [comment]
Awesome! It looks to be working well now. I am getting parameter estimates that are close to the true values for several simulated datasets.

Thank you for looking into this and solving the error. I do not regularly code in C++ and did not realize that `std::abs` is for integers. Explains the weird results I was getting.

**Comment by bbolker** [comment]
Please post a pull request with the new stuff (we also need tests and an entry in the `NEWS` file ...)

**Comment by psmorris15** [comment]
Pull request has been submitted. Please let me know what I can do to wrap those loose ends up.

**Comment by bbolker** [comment]
I added [tests and NEWS](https://github.com/glmmTMB/glmmTMB/pull/1042).  Remaining loose ends:

* your full name (or internet alias) plus ORCid ID (if desired) for the contributors list
* still need to improve printing of summary etc. (and `sigma()` printing, which is suboptimal for many families
* deviance residuals?
* it would be nice to figure out more reliable starting conditions. Using `log(sd(response_var))` seems to work OK for the dispersion component (or the intercept of the dispersion component, for a non-trivial dispersion model). The skew is harder, although [Wikipedia has a method-of-moments estimate](https://en.wikipedia.org/wiki/Skew_normal_distribution#Estimation) that would probably (?) give decent order-of-magnitude starting values if applied to the response variable? This gets into a different can of worms of [deciding on more robust starting values](https://github.com/glmmTMB/glmmTMB/issues/1036) ...

**Comment by psmorris15** [comment]
I'll continue to think about these issues when I have some time. 

ORCid ID is here: https://orcid.org/0000-0002-4035-0895

Thank you for your patience as this has come together.

**Comment by psmorris15** [comment]
Wanted to briefly follow up on a small thing that I noticed. I had originally referred to the shape parameter of the skew-normal distribution as "skew" in the documentation and print statement code. The correct language should be shape, as the parameter does not dictate the third moment directly. I had edited my code to reflect this change in wording but it looks like those changes were not merged. 

[](http://azzalini.stat.unipd.it/SN/Intro/)

**Comment by bbolker** [comment]
That rings a bell but I don't see where those changes are (there aren't obvious unmerged pieces in any of the pull requests I can see).  Are you willing to create a new PR?

**Comment by psmorris15** [comment]
I think I've pinpointed the issue. It is still listed as "skew" in the manual file but not in the `family.R` script. It looks like the `roxygen2` process needs to be rerun?

[https://github.com/glmmTMB/glmmTMB/blob/edbeab20e4d0948275ad2b3555a74d1092ca62cb/glmmTMB/man/nbinom2.Rd#L97](https://github.com/glmmTMB/glmmTMB/blob/edbeab20e4d0948275ad2b3555a74d1092ca62cb/glmmTMB/man/nbinom2.Rd#L97)

**Comment by bbolker** [comment]
Should be fixed in ad6d06ab3 .
